src/ORM/DB.php

```diff
@@ -696,4 +712,91 @@ public static function alteration_message($message, $type = "")
     {
         self::get_schema()->alterationMessage($message, $type);
     }
+
+    /**
+     * Check if all tables and field columns for a class exist in the database.
+     *
+     * @param string $class
+     * @return boolean
+     */
+    public static function database_is_ready(string $class): bool
+    {
+        if (!is_subclass_of($class, DataObject::class)) {
+            throw new InvalidArgumentException("$class is not a subclass of " . DataObject::class);
+        }
+
+        // Used for unit tests
+        if (array_key_exists($class, self::$force_database_is_ready)) {
+            return self::$force_database_is_ready[$class];
+        }
+
+        // Don't check again if we already know the db is ready for this class.
+        if (!empty(self::$database_is_ready[$class])) {
+            return true;
+        }
+
+        // Check if all tables and fields required for the class exist in the database.
+        $requiredClasses = ClassInfo::dataClassesFor($class);
+        $schema = DataObject::getSchema();
+        foreach ($requiredClasses as $required) {
+            // Skip test classes, as not all test classes are scaffolded at once
+            if (is_a($required, TestOnly::class, true)) {
+                continue;
+            }
+
+            // if any of the tables aren't created in the database
+            $table = $schema->tableName($required);
+            if (!ClassInfo::hasTable($table)) {
+                return false;
+            }
+
+            // HACK: DataExtensions aren't applied until a class is instantiated for
+            // the first time, so create an instance here.
+            singleton($required);
+
+            // if any of the tables don't have all fields mapped as table columns
+            $dbFields = DB::field_list($table);
+            if (!$dbFields) {
+                return false;
+            }
+
+            $objFields = $schema->databaseFields($required, false);
+            $missingFields = array_diff_key($objFields, $dbFields);
+
+            if ($missingFields) {
+                return false;
+            }
+        }
+        self::$database_is_ready[$class] = true;
+
+        return true;
+    }
+
+    /**
+     * Resets the database_is_ready cache.
+     *
+     * @param string $class The specific class to be cleared.
+     * If not passed, the cache for all classes is cleared.
+     */
+    public static function clear_database_is_ready(string $class)
+    {
+        if ($class) {
+            unset(self::$database_is_ready[$class]);
```

Comment:
`unset` can take mutliple arguments, so these lines can be consolidated

Reply:
Fair point. But I prefer your idea of consolidating the two arrays in any case, so It'll just have the one array to unset from anyway.